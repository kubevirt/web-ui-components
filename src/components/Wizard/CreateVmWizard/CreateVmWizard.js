import React from 'react';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';
import { Wizard } from 'patternfly-react';

import BasicSettingsTab from './BasicSettingsTab';
import StorageTab from './StorageTab';
import ResultTab from './ResultTab';

import { createVm, createVmTemplate } from '../../../k8s/request';
import {
  POD_NETWORK,
  PROVISION_SOURCE_URL,
  PROVISION_SOURCE_CONTAINER,
  PROVISION_SOURCE_PXE,
} from '../../../constants';
import { NetworksTab } from './NetworksTab';
import { settingsValue } from '../../../k8s/selectors';
import {
  PROVISION_SOURCE_TYPE_KEY,
  USER_TEMPLATE_KEY,
  BASIC_SETTINGS_TAB_KEY,
  NETWORKS_TAB_KEY,
  STORAGE_TAB_KEY,
  RESULT_TAB_KEY,
  STORAGE_TYPE_CONTAINER,
  STORAGE_TYPE_DATAVOLUME,
  NETWORK_TYPE_POD,
} from './constants';
import {
  CREATE_VM,
  CREATE_VM_TEMPLATE,
  STEP_BASIC_SETTINGS,
  STEP_NETWORK,
  STEP_STORAGE,
  STEP_RESULT,
  NEXT,
} from './strings';
import { loadingWizardTab } from '../loadingWizardTab';
import {
  getUserTemplate,
  getTemplateStorages,
  getTemplateInterfaces,
  hasAutoAttachPodInterface,
} from '../../../utils/templates';

const LoadingBasicWizardTab = loadingWizardTab(BasicSettingsTab);
const LoadingStorageTab = loadingWizardTab(StorageTab);
const LoadingNetworksTab = loadingWizardTab(NetworksTab);

const getBasicSettingsValue = (stepData, key) => settingsValue(stepData[BASIC_SETTINGS_TAB_KEY].value, key);

const onUserTemplateChangedInStorageTab = (stepData, newUserTemplate) => {
  const withoutDiscardedTemplateStorage = stepData[STORAGE_TAB_KEY].value.filter(storage => !storage.templateStorage);

  const rows = [...withoutDiscardedTemplateStorage];

  if (newUserTemplate) {
    const templateStorages = getTemplateStorages(newUserTemplate).map(storage => ({
      templateStorage: storage,
    }));
    rows.push(...templateStorages);
  }

  return {
    ...stepData,
    [STORAGE_TAB_KEY]: {
      ...stepData[STORAGE_TAB_KEY],
      value: rows,
    },
  };
};

const onUserTemplateChangedInNetworksTab = (stepData, newUserTemplate) => {
  const withoutDiscardedTemplateNetworks = stepData[NETWORKS_TAB_KEY].value.filter(network => !network.templateNetwork);

  const rows = [...withoutDiscardedTemplateNetworks];
  if (!rows.find(row => row.rootNetwork)) {
    rows.push(podNetwork);
  }

  if (newUserTemplate) {
    const templateInterfaces = getTemplateInterfaces(newUserTemplate);
    const networks = templateInterfaces.map(i => ({
      templateNetwork: i,
    }));

    // do not add root interface if there already is one or autoAttachPodInterface is set to false
    if (networks.some(network => network.templateNetwork.network.pod) || !hasAutoAttachPodInterface(newUserTemplate)) {
      const index = findIndex(rows, network => network.rootNetwork);
      rows.splice(index, 1);
    }

    rows.push(...networks);
  }

  return {
    ...stepData,
    [NETWORKS_TAB_KEY]: {
      ...stepData[NETWORKS_TAB_KEY],
      value: rows,
    },
  };
};

const rootDisk = {
  rootStorage: {},
  name: 'rootdisk',
  isBootable: true,
};

export const rootContainerDisk = {
  ...rootDisk,
  storageType: STORAGE_TYPE_CONTAINER,
};

export const rootDataVolumeDisk = {
  ...rootDisk,
  storageType: STORAGE_TYPE_DATAVOLUME,
  size: 10,
};

const onImageSourceTypeChangedInStorageTab = stepData => {
  const withoutRootStorage = stepData[STORAGE_TAB_KEY].value.filter(storage => !storage.rootStorage);
  const rows = [...withoutRootStorage];
  const basicSettings = stepData[BASIC_SETTINGS_TAB_KEY].value;
  if (!(basicSettings[USER_TEMPLATE_KEY] && basicSettings[USER_TEMPLATE_KEY].value)) {
    let storage;
    const provisionSource = basicSettings[PROVISION_SOURCE_TYPE_KEY].value;
    switch (provisionSource) {
      case PROVISION_SOURCE_URL:
        storage = rootDataVolumeDisk;
        break;
      case PROVISION_SOURCE_CONTAINER:
        storage = rootContainerDisk;
        break;
      case PROVISION_SOURCE_PXE:
        break;
      default:
        // eslint-disable-next-line
        console.warn(`Unknown provision source ${provisionSource}`);
        break;
    }
    if (storage) {
      rows.push(storage);
    }
  }
  return {
    ...stepData,
    [STORAGE_TAB_KEY]: {
      ...stepData[STORAGE_TAB_KEY],
      value: rows,
    },
  };
};

const onUserTemplateChanged = (props, stepData) => {
  const userTemplateName = getBasicSettingsValue(stepData, USER_TEMPLATE_KEY);
  const userTemplate = userTemplateName ? getUserTemplate(props.templates, userTemplateName) : undefined;
  const newStepData = onUserTemplateChangedInStorageTab(stepData, userTemplate);
  return onUserTemplateChangedInNetworksTab(newStepData, userTemplate);
};

const onImageSourceTypeChanged = (props, stepData) => onImageSourceTypeChangedInStorageTab(stepData);

const podNetwork = {
  rootNetwork: {},
  id: 0,
  name: 'eth0',
  mac: '',
  network: POD_NETWORK,
  editable: true,
  edit: false,
  networkType: NETWORK_TYPE_POD,
};

export class CreateVmWizard extends React.Component {
  state = {
    activeStepIndex: 0,
    stepData: {
      [BASIC_SETTINGS_TAB_KEY]: {
        // Basic Settings
        value: {},
        valid: false,
      },
      [NETWORKS_TAB_KEY]: {
        value: [podNetwork],
        valid: true,
      },
      [STORAGE_TAB_KEY]: {
        value: [], // Storages
        valid: true, // empty Storages are valid
      },
      [RESULT_TAB_KEY]: {
        value: '',
        valid: null, // result of the request
      },
    },
  };

  getLastStepIndex = () => this.wizardStepsNewVM.length - 1;

  lastStepReached = () => this.state.activeStepIndex === this.getLastStepIndex();

  callbacks = [
    {
      field: USER_TEMPLATE_KEY,
      callback: onUserTemplateChanged,
    },
    {
      field: PROVISION_SOURCE_TYPE_KEY,
      callback: onImageSourceTypeChanged,
    },
  ];

  onStepDataChanged = (key, value, valid) => {
    this.setState((state, props) => {
      let stepData = { ...state.stepData };

      stepData[key] = {
        value,
        valid,
      };

      if (key === BASIC_SETTINGS_TAB_KEY) {
        this.callbacks.forEach(callback => {
          const oldValue = getBasicSettingsValue(state.stepData, callback.field);
          const newValue = getBasicSettingsValue(stepData, callback.field);
          if (oldValue !== newValue) {
            stepData = callback.callback(props, stepData);
          }
        });
      }

      return { stepData };
    });
  };

  finish() {
    const create = this.props.createTemplate
      ? createVmTemplate
      : (k8sCreate, templates, basicSettings, networks, storage) =>
          createVm(k8sCreate, templates, basicSettings, networks, storage, this.props.persistentVolumeClaims);
    create(
      this.props.k8sCreate,
      this.props.templates,
      this.state.stepData[BASIC_SETTINGS_TAB_KEY].value,
      this.state.stepData[NETWORKS_TAB_KEY].value,
      this.state.stepData[STORAGE_TAB_KEY].value
    )
      .then(results => {
        const vmResult = results[results.length - 1];
        return this.onStepDataChanged(RESULT_TAB_KEY, JSON.stringify(vmResult, null, 1), true);
      })
      .catch(error => this.onStepDataChanged(RESULT_TAB_KEY, error.message, false));
  }

  onStepChanged = newActiveStepIndex => {
    if (
      !this.lastStepReached() && // do not allow going back once last step is reached
      (newActiveStepIndex < this.state.activeStepIndex || // allow going back to past steps
        this.state.stepData[this.wizardStepsNewVM[newActiveStepIndex - 1].key].valid)
    ) {
      if (newActiveStepIndex === this.getLastStepIndex()) {
        this.finish();
      }
      this.setState({ activeStepIndex: newActiveStepIndex });
    }
  };

  wizardStepsNewVM = [
    {
      title: STEP_BASIC_SETTINGS,
      key: BASIC_SETTINGS_TAB_KEY,
      render: () => {
        const loadingData = {
          namespaces: this.props.namespaces,
          templates: this.props.templates,
        };
        return (
          <LoadingBasicWizardTab
            key="1"
            selectedNamespace={this.props.selectedNamespace}
            basicSettings={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value}
            onChange={(value, valid) => this.onStepDataChanged(BASIC_SETTINGS_TAB_KEY, value, valid)}
            loadingData={loadingData}
            createTemplate={this.props.createTemplate}
          />
        );
      },
    },
    {
      title: STEP_NETWORK,
      key: NETWORKS_TAB_KEY,
      render: () => {
        const loadingData = {
          networkConfigs: this.props.networkConfigs,
        };
        const sourceType = getBasicSettingsValue(this.state.stepData, PROVISION_SOURCE_TYPE_KEY);
        return (
          <LoadingNetworksTab
            onChange={(value, valid) => this.onStepDataChanged(NETWORKS_TAB_KEY, value, valid)}
            networkConfigs={this.props.networkConfigs}
            networks={this.state.stepData[NETWORKS_TAB_KEY].value || []}
            sourceType={sourceType}
            namespace={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value.namespace.value}
            loadingData={loadingData}
          />
        );
      },
    },
    {
      title: STEP_STORAGE,
      key: STORAGE_TAB_KEY,
      render: () => {
        const loadingData = {
          storageClasses: this.props.storageClasses,
          persistentVolumeClaims: this.props.persistentVolumeClaims,
        };
        const sourceType = getBasicSettingsValue(this.state.stepData, PROVISION_SOURCE_TYPE_KEY);
        return (
          <LoadingStorageTab
            initialStorages={this.state.stepData[STORAGE_TAB_KEY].value}
            onChange={(value, valid) => this.onStepDataChanged(STORAGE_TAB_KEY, value, valid)}
            units={this.props.units}
            sourceType={sourceType}
            namespace={this.state.stepData[BASIC_SETTINGS_TAB_KEY].value.namespace.value}
            loadingData={loadingData}
          />
        );
      },
    },
    {
      title: STEP_RESULT,
      key: RESULT_TAB_KEY,
      render: () => {
        const stepData = this.state.stepData[RESULT_TAB_KEY];
        return (
          <ResultTab result={stepData.value} success={stepData.valid} createTemplate={this.props.createTemplate} />
        );
      },
    },
  ];

  render() {
    const beforeLastStepReached = this.state.activeStepIndex === this.wizardStepsNewVM.length - 2;
    const lastStepReached = this.lastStepReached();

    const createVmText = this.props.createTemplate ? CREATE_VM_TEMPLATE : CREATE_VM;

    return (
      <Wizard.Pattern
        show
        onHide={this.props.onHide}
        steps={this.wizardStepsNewVM}
        activeStepIndex={this.state.activeStepIndex}
        onStepChanged={index => this.onStepChanged(index)}
        previousStepDisabled={lastStepReached}
        cancelButtonDisabled={lastStepReached}
        stepButtonsDisabled={lastStepReached}
        nextStepDisabled={!this.state.stepData[this.wizardStepsNewVM[this.state.activeStepIndex].key].valid}
        nextText={beforeLastStepReached ? createVmText : NEXT}
        title={createVmText}
      />
    );
  }
}

CreateVmWizard.defaultProps = {
  selectedNamespace: undefined,
  templates: undefined,
  namespaces: undefined,
  networkConfigs: undefined,
  persistentVolumeClaims: undefined,
  storageClasses: undefined,
  createTemplate: false,
};

CreateVmWizard.propTypes = {
  onHide: PropTypes.func.isRequired,
  templates: PropTypes.array,
  namespaces: PropTypes.array,
  selectedNamespace: PropTypes.object,
  k8sCreate: PropTypes.func.isRequired,
  networkConfigs: PropTypes.array,
  persistentVolumeClaims: PropTypes.array,
  storageClasses: PropTypes.array,
  units: PropTypes.object.isRequired,
  createTemplate: PropTypes.bool,
};
