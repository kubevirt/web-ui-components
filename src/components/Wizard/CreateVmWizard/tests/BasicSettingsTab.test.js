import React from 'react';
import { shallow, mount } from 'enzyme';
import { MenuItem } from 'patternfly-react';

import { BasicSettingsTab, getFormFields } from '../BasicSettingsTab';
import { namespaces } from '../fixtures/CreateVmWizard.fixture';
import { baseTemplates } from '../../../../k8s/mock_templates';
import { validBasicSettings } from '../fixtures/BasicSettingsTab.fixture';
import { DNS1123_START_ERROR } from '../../../../utils/strings';
import { getValidationObject } from '../../../../utils/validations';
import { getName, getMemory, getCpu, getCloudInitUserData } from '../../../../utils/selectors';
import { getTemplateProvisionSource } from '../../../../utils/templates';
import { Dropdown } from '../../../Form';
import { NO_TEMPLATE } from '../strings';
import { selectVm } from '../../../../k8s/selectors';
import {
  userTemplates,
  urlTemplate,
  containerTemplate,
  pxeTemplate,
  containerCloudTemplate,
  urlCustomFlavorTemplate,
} from '../../../../k8s/mock_user_templates';

import {
  NAME_KEY,
  USER_TEMPLATE_KEY,
  PROVISION_SOURCE_TYPE_KEY,
  IMAGE_URL_KEY,
  CONTAINER_IMAGE_KEY,
  CLOUD_INIT_KEY,
  USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY,
  CLOUD_INIT_CUSTOM_SCRIPT_KEY,
  CPU_KEY,
  FLAVOR_KEY,
  MEMORY_KEY,
} from '../constants';

import {
  PROVISION_SOURCE_PXE,
  PROVISION_SOURCE_CONTAINER,
  PROVISION_SOURCE_URL,
  CUSTOM_FLAVOR,
} from '../../../../constants';

const templates = [...baseTemplates, ...userTemplates];

const testBasicSettingsTab = (basicSettings = {}, onChange = null, selectedNamespace = undefined, template = false) => (
  <BasicSettingsTab
    templates={templates}
    namespaces={namespaces}
    selectedNamespace={selectedNamespace}
    basicSettings={basicSettings}
    onChange={onChange || jest.fn()}
    createVmTemplate={template}
  />
);

const expectMockToBeCalledWith = (fn, a, b, call = 0) => {
  expect(fn.mock.calls[call][0]).toEqual(a);
  expect(fn.mock.calls[call][1]).toEqual(b);
};

const testFormChange = (what, value, result, valid) => {
  const onChange = jest.fn();
  const component = mount(testBasicSettingsTab({}, onChange));

  onFormChange(component, value, what);

  expectMockToBeCalledWith(onChange, result, valid);
};

const onFormChange = (component, value, what) => {
  const formFields = getFormFields(validBasicSettings, namespaces, templates);

  const field = component.find(`#${formFields[what].id}`);
  if (formFields[what].type === 'dropdown') {
    field
      .find(MenuItem)
      .findWhere(item => item.text() === value)
      .find('a')
      .simulate('click');
  } else {
    field.find('input').simulate('change', { target: { value } });
  }
};

const checkDisabledDropdown = (component, template, disabled) => {
  component.setProps({
    basicSettings: {
      ...component.props().basicSettings,
      [USER_TEMPLATE_KEY]: {
        value: getName(template),
      },
    },
  });

  expect(
    component
      .find({ id: 'image-source-type-dropdown' })
      .find(Dropdown)
      .props().disabled
  ).toEqual(disabled.image);
  expect(
    component
      .find({ id: 'operating-system-dropdown' })
      .find(Dropdown)
      .props().disabled
  ).toEqual(disabled.os);
  expect(
    component
      .find({ id: 'flavor-dropdown' })
      .find(Dropdown)
      .props().disabled
  ).toEqual(disabled.flavor);
  expect(
    component
      .find({ id: 'workload-profile-dropdown' })
      .find(Dropdown)
      .props().disabled
  ).toEqual(disabled.workload);
};

const clickOnUserTemplate = (component, templateName) => {
  const templatesDropdown = component.find('#template-dropdown').find(Dropdown);
  const userTemplateItem = templatesDropdown
    .find(MenuItem)
    .findWhere(i => i.text() === templateName)
    .find(MenuItem);
  expect(userTemplateItem.exists()).toBeTruthy();

  userTemplateItem.find('a').simulate('click');
};

describe('<BasicSettingsTab />', () => {
  it('renders correctly', () => {
    const component = shallow(testBasicSettingsTab());
    expect(component).toMatchSnapshot();
  });

  it('defaults to selectedNamespace', () => {
    const onChange = jest.fn();
    const namespace = namespaces[1];
    shallow(testBasicSettingsTab({}, onChange, namespace));

    expectMockToBeCalledWith(
      onChange,
      {
        namespace: {
          validation: undefined,
          value: namespace.metadata.name,
        },
      },
      false
    );
  });

  it('changes selectedNamespace on prop change', () => {
    const onChange = jest.fn();
    const basicSettings = {};
    const selectedNamespace = namespaces[1];

    const component = shallow(testBasicSettingsTab(basicSettings, onChange));

    expect(onChange).not.toHaveBeenCalled();

    component.setProps({
      templates,
      namespaces,
      selectedNamespace,
      basicSettings,
      onChange,
    });

    expectMockToBeCalledWith(
      onChange,
      {
        namespace: {
          validation: undefined,
          value: selectedNamespace.metadata.name,
        },
      },
      false
    );
  });

  it('validates incomplete form', () => {
    testFormChange(
      'name',
      'somename',
      {
        name: {
          validation: null,
          value: 'somename',
        },
      },
      false
    );
  });

  it('validates name field', () => {
    testFormChange(
      'name',
      'somename',
      {
        name: {
          validation: null,
          value: 'somename',
        },
      },
      false
    );
    testFormChange(
      'name',
      '_name',
      {
        name: {
          validation: getValidationObject(`Name ${DNS1123_START_ERROR}`),
          value: '_name',
        },
      },
      false
    );
  });

  it('is valid when all required fields are filled', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));
    onFormChange(component, validBasicSettings.name.value, 'name'); // trigger validation

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        flavor: {
          value: 'small',
          validation: undefined,
        },
        name: {
          ...validBasicSettings.name,
          validation: null,
        },
      },
      true
    );
  });

  it('required property is validated', () => {
    testFormChange(
      'name',
      '',
      {
        name: {
          validation: getValidationObject('Name is required'),
          value: '',
        },
      },
      false
    );
  });

  it('is invalid when one required fields is missing', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));

    onFormChange(component, '', NAME_KEY);

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        flavor: {
          value: 'small',
          validation: undefined,
        },
        name: {
          validation: getValidationObject('Name is required'),
          value: '',
        },
      },
      false
    );
  });

  it('reads provision source from user template', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));

    onFormChange(component, getName(urlTemplate), USER_TEMPLATE_KEY);
    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(urlTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_URL,
        },
        [IMAGE_URL_KEY]: {
          validation: undefined,
          value: getTemplateProvisionSource(urlTemplate).source,
        },
      },
      true
    );

    onFormChange(component, getName(containerTemplate), USER_TEMPLATE_KEY);
    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(containerTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_CONTAINER,
        },
        [CONTAINER_IMAGE_KEY]: {
          validation: undefined,
          value: getTemplateProvisionSource(containerTemplate).source,
        },
      },
      true,
      1
    );

    onFormChange(component, getName(pxeTemplate), USER_TEMPLATE_KEY);
    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(pxeTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_PXE,
        },
      },
      true,
      2
    );
  });

  it('reads cloud init settings from user template', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));

    onFormChange(component, getName(containerCloudTemplate), USER_TEMPLATE_KEY);
    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(containerCloudTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_CONTAINER,
        },
        [CONTAINER_IMAGE_KEY]: {
          value: 'fooContainer',
          validation: undefined,
        },
        [CLOUD_INIT_KEY]: {
          validation: undefined,
          value: true,
        },
        [USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
          validation: undefined,
          value: true,
        },
        [CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
          validation: undefined,
          value: getCloudInitUserData(selectVm(containerCloudTemplate.objects)),
        },
      },
      true
    );
  });

  it('reads flavor from user template', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));

    onFormChange(component, getName(containerCloudTemplate), USER_TEMPLATE_KEY);

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(containerCloudTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_CONTAINER,
        },
        [CONTAINER_IMAGE_KEY]: {
          value: getTemplateProvisionSource(containerCloudTemplate).source,
          validation: undefined,
        },
        [CLOUD_INIT_KEY]: {
          validation: undefined,
          value: true,
        },
        [USE_CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
          validation: undefined,
          value: true,
        },
        [CLOUD_INIT_CUSTOM_SCRIPT_KEY]: {
          validation: undefined,
          value: getCloudInitUserData(selectVm(containerCloudTemplate.objects)),
        },
      },
      true
    );

    onFormChange(component, getName(urlCustomFlavorTemplate), USER_TEMPLATE_KEY);

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(urlCustomFlavorTemplate),
          validation: undefined,
        },
        [PROVISION_SOURCE_TYPE_KEY]: {
          validation: undefined,
          value: PROVISION_SOURCE_URL,
        },
        [IMAGE_URL_KEY]: {
          value: getTemplateProvisionSource(urlCustomFlavorTemplate).source,
          validation: undefined,
        },
        [FLAVOR_KEY]: {
          value: CUSTOM_FLAVOR,
          validation: undefined,
        },
        [CPU_KEY]: {
          value: getCpu(urlCustomFlavorTemplate.objects[0]),
          validation: undefined,
        },
        [MEMORY_KEY]: {
          value: parseInt(getMemory(urlCustomFlavorTemplate.objects[0]), 10),
          validation: undefined,
        },
      },
      true,
      1
    );
  });

  it('provides no user template choice', () => {
    const onChange = jest.fn();
    const component = mount(testBasicSettingsTab(validBasicSettings, onChange));

    clickOnUserTemplate(component, getName(userTemplates[0]));

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: getName(userTemplates[0]),
          validation: undefined,
        },
        [CONTAINER_IMAGE_KEY]: {
          value: 'fooContainer',
        },
      },
      true
    );

    clickOnUserTemplate(component, NO_TEMPLATE);

    expectMockToBeCalledWith(
      onChange,
      {
        ...validBasicSettings,
        [USER_TEMPLATE_KEY]: {
          value: undefined,
          validation: undefined,
        },
      },
      true,
      1
    );
  });

  it('disables provision source, workload and os for user template', () => {
    const component = mount(testBasicSettingsTab(validBasicSettings));

    checkDisabledDropdown(component, containerTemplate, {
      image: true,
      os: true,
      flavor: false,
      workload: true,
    });

    checkDisabledDropdown(component, urlCustomFlavorTemplate, {
      image: true,
      os: true,
      flavor: true,
      workload: true,
    });

    checkDisabledDropdown(component, undefined, {
      image: false,
      os: false,
      flavor: false,
      workload: false,
    });
  });
});

describe('<BasicSettingsTab /> for Create VM Template', () => {
  it('renders correctly', () => {
    const component = shallow(testBasicSettingsTab({}, null, null, true));
    expect(component).toMatchSnapshot();
  });
});
