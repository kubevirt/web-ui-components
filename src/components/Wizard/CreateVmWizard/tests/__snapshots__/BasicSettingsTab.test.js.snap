// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasicSettingsTab /> for Create VM Template renders correctly 1`] = `
<FormFactory
  controlSize={5}
  fields={
    Object {
      "authKeys": Object {
        "id": "cloud-init-ssh",
        "isVisible": [Function],
        "title": "Authenticated SSH Keys",
        "type": "textarea",
      },
      "cloudInit": Object {
        "id": "use-cloud-init",
        "title": "Use cloud-init",
        "type": "checkbox",
      },
      "cpu": Object {
        "id": "resources-cpu",
        "isVisible": [Function],
        "required": true,
        "title": "CPUs",
        "type": "positive-number",
      },
      "description": Object {
        "id": "vm-description",
        "title": "Description",
        "type": "textarea",
      },
      "flavor": Object {
        "choices": Array [
          "small",
          "medium",
          "Custom",
        ],
        "defaultValue": "--- Select Flavor ---",
        "disabled": false,
        "id": "flavor-dropdown",
        "required": true,
        "title": "Flavor",
        "type": "dropdown",
      },
      "hostname": Object {
        "id": "cloud-init-hostname",
        "isVisible": [Function],
        "title": "Hostname",
      },
      "imageURL": Object {
        "disabled": false,
        "id": "provision-source-url",
        "isVisible": [Function],
        "required": true,
        "title": "URL",
      },
      "memory": Object {
        "id": "resources-memory",
        "isVisible": [Function],
        "required": true,
        "title": "Memory (GB)",
        "type": "positive-number",
      },
      "name": Object {
        "id": "vm-name",
        "required": true,
        "title": "Name",
        "validate": [Function],
      },
      "namespace": Object {
        "choices": Array [
          "default",
          "myproject",
        ],
        "defaultValue": "--- Select Namespace ---",
        "id": "namespace-dropdown",
        "required": true,
        "title": "Namespace",
        "type": "dropdown",
      },
      "operatingSystem": Object {
        "choices": Array [
          "fedora29",
          "fedora28",
          "fedora27",
          "fedora26",
          "fedora25",
          "fedora24",
          "fedora23",
          "rhel7.0",
          "ubuntu18.04",
          "win2k12r2",
          "win2k8r2",
          "win2k8",
          "win10",
        ],
        "defaultValue": "--- Select Operating System ---",
        "disabled": false,
        "id": "operating-system-dropdown",
        "required": true,
        "title": "Operating System",
        "type": "dropdown",
      },
      "provisionSourceType": Object {
        "choices": Array [
          "PXE",
          "URL",
          "Registry",
        ],
        "defaultValue": "--- Select Provision Source ---",
        "disabled": false,
        "id": "image-source-type-dropdown",
        "required": true,
        "title": "Provision Source",
        "type": "dropdown",
      },
      "registryImage": Object {
        "disabled": false,
        "id": "provision-source-registry",
        "isVisible": [Function],
        "required": true,
        "title": "Registry Image",
      },
      "startVM": Object {
        "id": "start-vm",
        "noBottom": true,
        "title": "Start virtual machine on creation",
        "type": "checkbox",
      },
      "userTemplate": Object {
        "choices": Array [
          "container-template",
          "container-template-cloud-init",
          "container-template-multus",
          "pxe-template",
          "pxe-template-dv",
          "pxe-template-pvc",
          "url-template",
          "url-template-custom-flavor",
          "url-template-nonetwork",
          "--- No Template ---",
        ],
        "defaultValue": "--- Select Template ---",
        "id": "template-dropdown",
        "title": "Template",
        "type": "dropdown",
      },
      "workloadProfile": Object {
        "choices": Array [
          "generic",
          "high-performance",
        ],
        "defaultValue": "--- Select Workload Profile ---",
        "disabled": false,
        "help": Array [
          <p>
            <b>
              generic
            </b>
            : 
            generic
          </p>,
          <p>
            <b>
              high-performance
            </b>
            : 
            high-performance
          </p>,
        ],
        "id": "workload-profile-dropdown",
        "required": true,
        "title": "Workload Profile",
        "type": "dropdown",
      },
    }
  }
  fieldsValues={Object {}}
  labelSize={3}
  onFormChange={[Function]}
  textPosition="text-right"
/>
`;

exports[`<BasicSettingsTab /> renders correctly 1`] = `
<FormFactory
  controlSize={5}
  fields={
    Object {
      "authKeys": Object {
        "id": "cloud-init-ssh",
        "isVisible": [Function],
        "title": "Authenticated SSH Keys",
        "type": "textarea",
      },
      "cloudInit": Object {
        "id": "use-cloud-init",
        "title": "Use cloud-init",
        "type": "checkbox",
      },
      "cpu": Object {
        "id": "resources-cpu",
        "isVisible": [Function],
        "required": true,
        "title": "CPUs",
        "type": "positive-number",
      },
      "description": Object {
        "id": "vm-description",
        "title": "Description",
        "type": "textarea",
      },
      "flavor": Object {
        "choices": Array [
          "small",
          "medium",
          "Custom",
        ],
        "defaultValue": "--- Select Flavor ---",
        "disabled": false,
        "id": "flavor-dropdown",
        "required": true,
        "title": "Flavor",
        "type": "dropdown",
      },
      "hostname": Object {
        "id": "cloud-init-hostname",
        "isVisible": [Function],
        "title": "Hostname",
      },
      "imageURL": Object {
        "disabled": false,
        "id": "provision-source-url",
        "isVisible": [Function],
        "required": true,
        "title": "URL",
      },
      "memory": Object {
        "id": "resources-memory",
        "isVisible": [Function],
        "required": true,
        "title": "Memory (GB)",
        "type": "positive-number",
      },
      "name": Object {
        "id": "vm-name",
        "required": true,
        "title": "Name",
        "validate": [Function],
      },
      "namespace": Object {
        "choices": Array [
          "default",
          "myproject",
        ],
        "defaultValue": "--- Select Namespace ---",
        "id": "namespace-dropdown",
        "required": true,
        "title": "Namespace",
        "type": "dropdown",
      },
      "operatingSystem": Object {
        "choices": Array [
          "fedora29",
          "fedora28",
          "fedora27",
          "fedora26",
          "fedora25",
          "fedora24",
          "fedora23",
          "rhel7.0",
          "ubuntu18.04",
          "win2k12r2",
          "win2k8r2",
          "win2k8",
          "win10",
        ],
        "defaultValue": "--- Select Operating System ---",
        "disabled": false,
        "id": "operating-system-dropdown",
        "required": true,
        "title": "Operating System",
        "type": "dropdown",
      },
      "provisionSourceType": Object {
        "choices": Array [
          "PXE",
          "URL",
          "Registry",
        ],
        "defaultValue": "--- Select Provision Source ---",
        "disabled": false,
        "id": "image-source-type-dropdown",
        "required": true,
        "title": "Provision Source",
        "type": "dropdown",
      },
      "registryImage": Object {
        "disabled": false,
        "id": "provision-source-registry",
        "isVisible": [Function],
        "required": true,
        "title": "Registry Image",
      },
      "startVM": Object {
        "id": "start-vm",
        "noBottom": true,
        "title": "Start virtual machine on creation",
        "type": "checkbox",
      },
      "userTemplate": Object {
        "choices": Array [
          "container-template",
          "container-template-cloud-init",
          "container-template-multus",
          "pxe-template",
          "pxe-template-dv",
          "pxe-template-pvc",
          "url-template",
          "url-template-custom-flavor",
          "url-template-nonetwork",
          "--- No Template ---",
        ],
        "defaultValue": "--- Select Template ---",
        "id": "template-dropdown",
        "title": "Template",
        "type": "dropdown",
      },
      "workloadProfile": Object {
        "choices": Array [
          "generic",
          "high-performance",
        ],
        "defaultValue": "--- Select Workload Profile ---",
        "disabled": false,
        "help": Array [
          <p>
            <b>
              generic
            </b>
            : 
            generic
          </p>,
          <p>
            <b>
              high-performance
            </b>
            : 
            high-performance
          </p>,
        ],
        "id": "workload-profile-dropdown",
        "required": true,
        "title": "Workload Profile",
        "type": "dropdown",
      },
    }
  }
  fieldsValues={Object {}}
  labelSize={3}
  onFormChange={[Function]}
  textPosition="text-right"
/>
`;
